url79d4f7c7ed_schanq_Live_Algorithm_tgz-pJ8-algorithms_and_systems_AI_Evolutionary_GeneticAlgorithmJ8/src/algorithms_and_systems_AI_Evolutionary/FuzzySystem.java

```diff
@@ -0,0 +1,938 @@
+	// #			determining Tonality, desiredLoudness, feel #
+	// #			for fitness function						#
+	// ##########################################################
+	
+	
+	// ==========================================================
+	// =				DEFUZZIFICATION METHODS					=
+	// ==========================================================
+
+	// Used in the Maximum Defuzzification technique			
+	public void sort(double [] set){
+		
+		Arrays.sort(set);
+		
+	}
+	public void sort(int [] set){
+		
+		Arrays.sort(set);
+		
+	}
+	// The Maximum Defuzzification Technique: хA(x*) │ хA(x)	
+	// for all x E X											
+	public double defuzz_Max(double [] u, double [] x){
+	
+		// Return error if the degree array (u) and 
+		// crisp array(x) are not the same length
+		if(u.length != x.length){
+			System.out.println("YOUR DEGREE INPUTS AND DESIRED OUTPUT ARRAYS ARE NOT EQUAL IN LENGTH");
+			return -99999;
+		}
+		
+		double crisp = 0;
+		double [] max = u.clone();
+		
+		this.sort(max);
+		
+		// If the maximum degree - marginalValue (deduct a small value to check 
+		// if it really is a peak value: set to 0.1) is still greater than the
+		// next highest value, set the crisp output to this value, else set 
+		// crisp to -1 so the singleton method may be used instead.
+		if(max[max.length-1]-0.1 > max[max.length-2]){
+			for(int i = 0; i < u.length; i++){
+				if(u[i] == max[max.length-1]){
+					crisp = x[i];
+					break;
+				}
+			}
+		}
+		else{
+			crisp = -1;
+		}
+		return crisp;
+		
+	}
+	// Weighted Average Defuzzification Technique or Singleton  
+	// technique:  хA(x*) = (ихixi)/(иxi)						
+	public double defuzz_Singleton(double [] u, double [] x){
+		
+		// Return error if the degree array (u) and 
+		// crisp array(x) are not the same length
+		if(u.length != x.length){
+			System.out.println("YOUR DEGREE INPUTS AND DESIRED OUTPUT ARRAYS ARE NOT EQUAL IN LENGTH");
+			return -99999;
+		}
+		
+		double sumOfDegrees = 0;
+		double sumOfDegreesMultipliedByCrisps = 0;
+		
+		for(int i = 0; i < u.length; i++)	
+			sumOfDegrees += u[i];
+		
+		for(int i = 0; i < x.length; i++)
+			sumOfDegreesMultipliedByCrisps += u[i]*x[i];
+		
+		return sumOfDegreesMultipliedByCrisps/sumOfDegrees;
+			
+	}
+	
+	// ==========================================================
+	// =		METHOD TO DETERMINE TARGET TONALITY				=
+	// ==========================================================
+	// Method to determine the Tonality that is most true		=
+	// for the performers current playing state. The 			=
+	// output to be fed into the fitness function of the Genetic= 
+	// Algorithm.												=
+	// NOTE: This method is not a fuzzy assessment: it decides 	=
+	// by comparing the input notes with the major scale of 	=
+	// each key, the closest matching (ie the highest occuring) =
+	// will determine the tonality.								=
+	// ==========================================================
+
+	private Tonality determineTonality(int [] notes){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 35. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 340-356 into a separate method.